Breaking change to `Generics::lifetimes` in v2.0.73

The following (contrived extraction from `der_derive`, see https://github.com/RustCrypto/formats/issues/1471) compiles with `syn` v2.0.72 but not with v2.0.73:

```rust
use syn::{Generics, GenericParam, Lifetime, LifetimeParam};
use proc_macro2::Span;

pub fn process_generics(mut generics: Generics) {
    let _lifetime = generics
        .lifetimes()
        .next()
        .map(|lt| lt.lifetime.clone())
        .unwrap_or_else(|| {
            let lt = Lifetime::new("'__default", Span::call_site());
            generics
                .params
                .insert(0, GenericParam::Lifetime(LifetimeParam::new(lt.clone())));
            lt
        });
}
```

It looks like it now assumes there could be a `Drop` impl where before it knew the concrete `Lifetimes` type did not have one:

```
error[E0502]: cannot borrow `generics.params` as mutable because it is also borrowed as immutable
  --> src/main.rs:9:25
   |
5  |       let _lifetime = generics
   |                       --------
   |                       |
   |  _____________________immutable borrow occurs here
   | |
6  | |         .lifetimes()
   | |____________________- a temporary with access to the immutable borrow is created here ...
...
9  |           .unwrap_or_else(|| {
   |                           ^^ mutable borrow occurs here
10 |               let lt = Lifetime::new("'__default", Span::call_site());
11 | /             generics
12 | |                 .params
   | |_______________________- second borrow occurs due to use of `generics.params` in closure
...
15 |           });
   |             - ... and the immutable borrow might be used here, when that temporary is dropped and runs the destructor for type `impl Iterator<Item = &LifetimeParam>`
```

The culprit seems to be ac9e1dd.

--- a/src/data.rs
+++ b/src/data.rs
@@ -138,43 +138,6 @@ impl Fields {
     /// self.a.clone() }`. For structs with unnamed fields, `Self { 0:
     /// self.0.clone() }`. And for unit structs, `Self {}`.
     pub fn members(&self) -> impl Iterator<Item = Member> + Clone + '_ {
-        struct Members<'a> {
-            fields: punctuated::Iter<'a, Field>,
-            index: u32,
-        }
-
-        impl<'a> Iterator for Members<'a> {
-            type Item = Member;
-
-            fn next(&mut self) -> Option<Self::Item> {
-                let field = self.fields.next()?;
-                let member = match &field.ident {
-                    Some(ident) => Member::Named(ident.clone()),
-                    None => {
-                        #[cfg(all(feature = "parsing", feature = "printing"))]
-                        let span = crate::spanned::Spanned::span(&field.ty);
-                        #[cfg(not(all(feature = "parsing", feature = "printing")))]
-                        let span = proc_macro2::Span::call_site();
-                        Member::Unnamed(Index {
-                            index: self.index,
-                            span,
-                        })
-                    }
-                };
-                self.index += 1;
-                Some(member)
-            }
-        }
-
-        impl<'a> Clone for Members<'a> {
-            fn clone(&self) -> Self {
-                Members {
-                    fields: self.fields.clone(),
-                    index: self.index,
-                }
-            }
-        }
-
         Members {
             fields: self.iter(),
             index: 0,
@@ -231,6 +194,43 @@ ast_struct! {
         pub colon_token: Option<Token![:]>,
 
         pub ty: Type,
+    }
+}
+
+pub struct Members<'a> {
+    fields: punctuated::Iter<'a, Field>,
+    index: u32,
+}
+
+impl<'a> Iterator for Members<'a> {
+    type Item = Member;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let field = self.fields.next()?;
+        let member = match &field.ident {
+            Some(ident) => Member::Named(ident.clone()),
+            None => {
+                #[cfg(all(feature = "parsing", feature = "printing"))]
+                let span = crate::spanned::Spanned::span(&field.ty);
+                #[cfg(not(all(feature = "parsing", feature = "printing")))]
+                let span = proc_macro2::Span::call_site();
+                Member::Unnamed(Index {
+                    index: self.index,
+                    span,
+                })
+            }
+        };
+        self.index += 1;
+        Some(member)
+    }
+}
+
+impl<'a> Clone for Members<'a> {
+    fn clone(&self) -> Self {
+        Members {
+            fields: self.fields.clone(),
+            index: self.index,
+        }
     }
 }
 

--- a/src/data.rs
+++ b/src/data.rs
@@ -105,42 +105,44 @@ impl Fields {
         }
     }
 
-    /// Get an iterator over the fields of a struct or variant as [`Member`]s.
-    /// This iterator can be used to iterate over a named or unnamed struct or
-    /// variant's fields uniformly.
-    ///
-    /// # Example
-    ///
-    /// The following is a simplistic [`Clone`] derive for structs. (A more
-    /// complete implementation would additionally want to infer trait bounds on
-    /// the generic type parameters.)
-    ///
-    /// ```
-    /// # use quote::quote;
-    /// #
-    /// fn derive_clone(input: &syn::ItemStruct) -> proc_macro2::TokenStream {
-    ///     let ident = &input.ident;
-    ///     let members = input.fields.members();
-    ///     let (impl_generics, ty_generics, where_clause) = input.generics.split_for_impl();
-    ///     quote! {
-    ///         impl #impl_generics Clone for #ident #ty_generics #where_clause {
-    ///             fn clone(&self) -> Self {
-    ///                 Self {
-    ///                     #(#members: self.#members.clone()),*
-    ///                 }
-    ///             }
-    ///         }
-    ///     }
-    /// }
-    /// ```
-    ///
-    /// For structs with named fields, it produces an expression like `Self { a:
-    /// self.a.clone() }`. For structs with unnamed fields, `Self { 0:
-    /// self.0.clone() }`. And for unit structs, `Self {}`.
-    pub fn members(&self) -> impl Iterator<Item = Member> + Clone + '_ {
-        Members {
-            fields: self.iter(),
-            index: 0,
+    return_impl_trait! {
+        /// Get an iterator over the fields of a struct or variant as [`Member`]s.
+        /// This iterator can be used to iterate over a named or unnamed struct or
+        /// variant's fields uniformly.
+        ///
+        /// # Example
+        ///
+        /// The following is a simplistic [`Clone`] derive for structs. (A more
+        /// complete implementation would additionally want to infer trait bounds on
+        /// the generic type parameters.)
+        ///
+        /// ```
+        /// # use quote::quote;
+        /// #
+        /// fn derive_clone(input: &syn::ItemStruct) -> proc_macro2::TokenStream {
+        ///     let ident = &input.ident;
+        ///     let members = input.fields.members();
+        ///     let (impl_generics, ty_generics, where_clause) = input.generics.split_for_impl();
+        ///     quote! {
+        ///         impl #impl_generics Clone for #ident #ty_generics #where_clause {
+        ///             fn clone(&self) -> Self {
+        ///                 Self {
+        ///                     #(#members: self.#members.clone()),*
+        ///                 }
+        ///             }
+        ///         }
+        ///     }
+        /// }
+        /// ```
+        ///
+        /// For structs with named fields, it produces an expression like `Self { a:
+        /// self.a.clone() }`. For structs with unnamed fields, `Self { 0:
+        /// self.0.clone() }`. And for unit structs, `Self {}`.
+        pub fn members(&self) -> impl Iterator<Item = Member> + Clone + '_ [Members] {
+            Members {
+                fields: self.iter(),
+                index: 0,
+            }
         }
     }
 }

--- a/src/generics.rs
+++ b/src/generics.rs
@@ -103,34 +103,46 @@ impl Default for Generics {
 }
 
 impl Generics {
-    /// Iterator over the lifetime parameters in `self.params`.
-    pub fn lifetimes(&self) -> impl Iterator<Item = &LifetimeParam> {
-        Lifetimes(self.params.iter())
+    return_impl_trait! {
+        /// Iterator over the lifetime parameters in `self.params`.
+        pub fn lifetimes(&self) -> impl Iterator<Item = &LifetimeParam> [Lifetimes] {
+            Lifetimes(self.params.iter())
+        }
     }
 
-    /// Iterator over the lifetime parameters in `self.params`.
-    pub fn lifetimes_mut(&mut self) -> impl Iterator<Item = &mut LifetimeParam> {
-        LifetimesMut(self.params.iter_mut())
+    return_impl_trait! {
+        /// Iterator over the lifetime parameters in `self.params`.
+        pub fn lifetimes_mut(&mut self) -> impl Iterator<Item = &mut LifetimeParam> [LifetimesMut] {
+            LifetimesMut(self.params.iter_mut())
+        }
     }
 
-    /// Iterator over the type parameters in `self.params`.
-    pub fn type_params(&self) -> impl Iterator<Item = &TypeParam> {
-        TypeParams(self.params.iter())
+    return_impl_trait! {
+        /// Iterator over the type parameters in `self.params`.
+        pub fn type_params(&self) -> impl Iterator<Item = &TypeParam> [TypeParams] {
+            TypeParams(self.params.iter())
+        }
     }
 
-    /// Iterator over the type parameters in `self.params`.
-    pub fn type_params_mut(&mut self) -> impl Iterator<Item = &mut TypeParam> {
-        TypeParamsMut(self.params.iter_mut())
+    return_impl_trait! {
+        /// Iterator over the type parameters in `self.params`.
+        pub fn type_params_mut(&mut self) -> impl Iterator<Item = &mut TypeParam> [TypeParamsMut] {
+            TypeParamsMut(self.params.iter_mut())
+        }
     }
 
-    /// Iterator over the constant parameters in `self.params`.
-    pub fn const_params(&self) -> impl Iterator<Item = &ConstParam> {
-        ConstParams(self.params.iter())
+    return_impl_trait! {
+        /// Iterator over the constant parameters in `self.params`.
+        pub fn const_params(&self) -> impl Iterator<Item = &ConstParam> [ConstParams] {
+            ConstParams(self.params.iter())
+        }
     }
 
-    /// Iterator over the constant parameters in `self.params`.
-    pub fn const_params_mut(&mut self) -> impl Iterator<Item = &mut ConstParam> {
-        ConstParamsMut(self.params.iter_mut())
+    return_impl_trait! {
+        /// Iterator over the constant parameters in `self.params`.
+        pub fn const_params_mut(&mut self) -> impl Iterator<Item = &mut ConstParam> [ConstParamsMut] {
+            ConstParamsMut(self.params.iter_mut())
+        }
     }
 
     /// Initializes an empty `where`-clause if there is not one present already.

--- a/src/generics.rs
+++ b/src/generics.rs
@@ -105,139 +105,31 @@ impl Default for Generics {
 impl Generics {
     /// Iterator over the lifetime parameters in `self.params`.
     pub fn lifetimes(&self) -> impl Iterator<Item = &LifetimeParam> {
-        struct Lifetimes<'a>(Iter<'a, GenericParam>);
-
-        impl<'a> Iterator for Lifetimes<'a> {
-            type Item = &'a LifetimeParam;
-
-            fn next(&mut self) -> Option<Self::Item> {
-                let next = match self.0.next() {
-                    Some(item) => item,
-                    None => return None,
-                };
-                if let GenericParam::Lifetime(lifetime) = next {
-                    Some(lifetime)
-                } else {
-                    self.next()
-                }
-            }
-        }
-
         Lifetimes(self.params.iter())
     }
 
     /// Iterator over the lifetime parameters in `self.params`.
     pub fn lifetimes_mut(&mut self) -> impl Iterator<Item = &mut LifetimeParam> {
-        struct LifetimesMut<'a>(IterMut<'a, GenericParam>);
-
-        impl<'a> Iterator for LifetimesMut<'a> {
-            type Item = &'a mut LifetimeParam;
-
-            fn next(&mut self) -> Option<Self::Item> {
-                let next = match self.0.next() {
-                    Some(item) => item,
-                    None => return None,
-                };
-                if let GenericParam::Lifetime(lifetime) = next {
-                    Some(lifetime)
-                } else {
-                    self.next()
-                }
-            }
-        }
-
         LifetimesMut(self.params.iter_mut())
     }
 
     /// Iterator over the type parameters in `self.params`.
     pub fn type_params(&self) -> impl Iterator<Item = &TypeParam> {
-        struct TypeParams<'a>(Iter<'a, GenericParam>);
-
-        impl<'a> Iterator for TypeParams<'a> {
-            type Item = &'a TypeParam;
-
-            fn next(&mut self) -> Option<Self::Item> {
-                let next = match self.0.next() {
-                    Some(item) => item,
-                    None => return None,
-                };
-                if let GenericParam::Type(type_param) = next {
-                    Some(type_param)
-                } else {
-                    self.next()
-                }
-            }
-        }
-
         TypeParams(self.params.iter())
     }
 
     /// Iterator over the type parameters in `self.params`.
     pub fn type_params_mut(&mut self) -> impl Iterator<Item = &mut TypeParam> {
-        struct TypeParamsMut<'a>(IterMut<'a, GenericParam>);
-
-        impl<'a> Iterator for TypeParamsMut<'a> {
-            type Item = &'a mut TypeParam;
-
-            fn next(&mut self) -> Option<Self::Item> {
-                let next = match self.0.next() {
-                    Some(item) => item,
-                    None => return None,
-                };
-                if let GenericParam::Type(type_param) = next {
-                    Some(type_param)
-                } else {
-                    self.next()
-                }
-            }
-        }
-
         TypeParamsMut(self.params.iter_mut())
     }
 
     /// Iterator over the constant parameters in `self.params`.
     pub fn const_params(&self) -> impl Iterator<Item = &ConstParam> {
-        struct ConstParams<'a>(Iter<'a, GenericParam>);
-
-        impl<'a> Iterator for ConstParams<'a> {
-            type Item = &'a ConstParam;
-
-            fn next(&mut self) -> Option<Self::Item> {
-                let next = match self.0.next() {
-                    Some(item) => item,
-                    None => return None,
-                };
-                if let GenericParam::Const(const_param) = next {
-                    Some(const_param)
-                } else {
-                    self.next()
-                }
-            }
-        }
-
         ConstParams(self.params.iter())
     }
 
     /// Iterator over the constant parameters in `self.params`.
     pub fn const_params_mut(&mut self) -> impl Iterator<Item = &mut ConstParam> {
-        struct ConstParamsMut<'a>(IterMut<'a, GenericParam>);
-
-        impl<'a> Iterator for ConstParamsMut<'a> {
-            type Item = &'a mut ConstParam;
-
-            fn next(&mut self) -> Option<Self::Item> {
-                let next = match self.0.next() {
-                    Some(item) => item,
-                    None => return None,
-                };
-                if let GenericParam::Const(const_param) = next {
-                    Some(const_param)
-                } else {
-                    self.next()
-                }
-            }
-        }
-
         ConstParamsMut(self.params.iter_mut())
     }
 
@@ -275,6 +167,114 @@ impl Generics {
             TypeGenerics(self),
             self.where_clause.as_ref(),
         )
+    }
+}
+
+pub struct Lifetimes<'a>(Iter<'a, GenericParam>);
+
+impl<'a> Iterator for Lifetimes<'a> {
+    type Item = &'a LifetimeParam;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let next = match self.0.next() {
+            Some(item) => item,
+            None => return None,
+        };
+        if let GenericParam::Lifetime(lifetime) = next {
+            Some(lifetime)
+        } else {
+            self.next()
+        }
+    }
+}
+
+pub struct LifetimesMut<'a>(IterMut<'a, GenericParam>);
+
+impl<'a> Iterator for LifetimesMut<'a> {
+    type Item = &'a mut LifetimeParam;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let next = match self.0.next() {
+            Some(item) => item,
+            None => return None,
+        };
+        if let GenericParam::Lifetime(lifetime) = next {
+            Some(lifetime)
+        } else {
+            self.next()
+        }
+    }
+}
+
+pub struct TypeParams<'a>(Iter<'a, GenericParam>);
+
+impl<'a> Iterator for TypeParams<'a> {
+    type Item = &'a TypeParam;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let next = match self.0.next() {
+            Some(item) => item,
+            None => return None,
+        };
+        if let GenericParam::Type(type_param) = next {
+            Some(type_param)
+        } else {
+            self.next()
+        }
+    }
+}
+
+pub struct TypeParamsMut<'a>(IterMut<'a, GenericParam>);
+
+impl<'a> Iterator for TypeParamsMut<'a> {
+    type Item = &'a mut TypeParam;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let next = match self.0.next() {
+            Some(item) => item,
+            None => return None,
+        };
+        if let GenericParam::Type(type_param) = next {
+            Some(type_param)
+        } else {
+            self.next()
+        }
+    }
+}
+
+pub struct ConstParams<'a>(Iter<'a, GenericParam>);
+
+impl<'a> Iterator for ConstParams<'a> {
+    type Item = &'a ConstParam;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let next = match self.0.next() {
+            Some(item) => item,
+            None => return None,
+        };
+        if let GenericParam::Const(const_param) = next {
+            Some(const_param)
+        } else {
+            self.next()
+        }
+    }
+}
+
+pub struct ConstParamsMut<'a>(IterMut<'a, GenericParam>);
+
+impl<'a> Iterator for ConstParamsMut<'a> {
+    type Item = &'a mut ConstParam;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let next = match self.0.next() {
+            Some(item) => item,
+            None => return None,
+        };
+        if let GenericParam::Const(const_param) = next {
+            Some(const_param)
+        } else {
+            self.next()
+        }
     }
 }
 

--- a/src/macros.rs
+++ b/src/macros.rs
@@ -164,3 +164,19 @@ macro_rules! check_keyword_matches {
     (pub pub) => {};
     (struct struct) => {};
 }
+
+#[cfg(any(feature = "full", feature = "derive"))]
+macro_rules! return_impl_trait {
+    (
+        $(#[$attr:meta])*
+        $vis:vis fn $name:ident $args:tt -> $impl_trait:ty [$concrete:ty] $body:block
+    ) => {
+        #[cfg(not(docsrs))]
+        $(#[$attr])*
+        $vis fn $name $args -> $concrete $body
+
+        #[cfg(docsrs)]
+        $(#[$attr])*
+        $vis fn $name $args -> $impl_trait $body
+    };
+}

--- a/tests/test_iterators.rs
+++ b/tests/test_iterators.rs
@@ -1,7 +1,7 @@
-#![allow(clippy::uninlined_format_args)]
+#![allow(clippy::map_unwrap_or, clippy::uninlined_format_args)]
 
 use syn::punctuated::{Pair, Punctuated};
-use syn::Token;
+use syn::{parse_quote, GenericParam, Generics, Lifetime, LifetimeParam, Token};
 
 #[macro_use]
 mod macros;
@@ -67,4 +67,23 @@ fn may_dangle() {
             break;
         }
     }
+}
+
+// Regression test for https://github.com/dtolnay/syn/issues/1718
+#[test]
+fn no_opaque_drop() {
+    let mut generics = Generics::default();
+
+    let _ = generics
+        .lifetimes()
+        .next()
+        .map(|param| param.lifetime.clone())
+        .unwrap_or_else(|| {
+            let lifetime: Lifetime = parse_quote!('a);
+            generics.params.insert(
+                0,
+                GenericParam::Lifetime(LifetimeParam::new(lifetime.clone())),
+            );
+            lifetime
+        });
 }

